lib/src/DynamicProto.ts

```diff
@@ -229,44 +250,96 @@ function _getBaseFuncs(classProto:any, thisTarget:any, instFuncs:any): any {
-            if (_this && _hasOwnProperty(proto, DynClassName)) {
-                let instFunc = ((_this[DynInstFuncTable] || {})[proto[DynClassName]] || {})[funcName];
-                if (instFunc) {
-                    // Used the instance function property
-                    return instFunc.apply(_this, arguments);
+function _getInstFunc(target: any, funcName: string, proto: any, currentDynProtoProxy: any) {
+    let instFunc = null;
+
+    // We need to check whether the class name is defined directly on this prototype otherwise
+    // it will walk the proto chain and return any parent proto classname.
+    if (target && _hasOwnProperty(proto, DynClassName)) {
+
+        let instFuncTable = target[DynInstFuncTable] || {};
+        instFunc = (instFuncTable[proto[DynClassName]] || {})[funcName];
+
+        if (!instFunc) {
+            // Avoid stack overflow from recursive calling the same function
+            _throwTypeError("Missing [" + funcName + "] " + strFunction);
+        }
+
+        // We have the instance function, lets check it we can speed up further calls
+        // by adding the instance function back directly on the instance (avoiding the dynamic func lookup)
+        if (!instFunc[DynInstChkTag] && instFuncTable[DynAllowInstChkTag] !== false) {
+            // If the instance already has an instance function we can't replace it
+            let canAddInst = !_hasOwnProperty(target, funcName);
+
+            // Get current prototype
+            let objProto = _getObjProto(target);
+            let visited:any[] = [];
+
+            // Lookup the function starting at the top (instance level prototype) and traverse down, if the first matching function
+            // if nothing is found or if the first hit is a dynamic proto instance then we can safely add an instance shortcut
+            while (canAddInst && objProto && !_isObjectArrayOrFunctionPrototype(objProto) && !_hasVisited(visited, objProto)) {
+                let protoFunc = objProto[funcName];
+                if (protoFunc) {
+                    canAddInst = (protoFunc === currentDynProtoProxy);
+                    break;
                 }
 
-                // Avoid stack overflow from recursive calling the same function
-                _throwTypeError("Missing [" + funcName + "] " + strFunction);
+                // We need to find all possible initial functions to ensure that we don't bypass a valid override function
+                visited.push(objProto);
+                objProto = _getObjProto(objProto);
             }
 
-            let protoFunc = proto[funcName];
+            try {
+                if (canAddInst) {
+                    // This instance doesn't have an instance func and the class hierarchy does have a higher level prototype version
+                    // so it's safe to directly assign for any subsequent calls (for better performance)
+                    target[funcName] = instFunc;
```

Comment:
This is where it assigns an instance level function to the looked up dynamic proto instance.